File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/P.Частичная_сортировка/solution.py
import sys


def solution(n: int, nums: list[int]) -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(1)
    """
    blocks: int = 0
    threshold: int = 0
    for idx in range(n):
        if nums[idx] > idx:
            threshold = max(threshold, nums[idx])
        if idx >= threshold:
            blocks += 1
            threshold = 0
    sys.stdout.write(str(blocks) + '\n')


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    solution(n, nums)

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/P.Частичная_сортировка/solution.cpp
#include <iostream>
#include <vector>


void solution(int& n, std::vector<int>& nums) {
    /*
    Time Complexity O(N)
    Memory Complexity O(1)
    */
    int blocks = 0;
    int start = 0;
    int threshold = 0;
    for (int idx = 0; idx < n; ++idx) {
        if (nums[idx] > idx) {
            threshold = std::max(threshold, nums[idx]);
        }
        if (idx >= threshold) {
            ++blocks;
            threshold = 0;
        }
    }
    std::cout << blocks << "\n";
}


int main() {
    int n;
    std::cin >> n;
    std::vector<int> nums;
    for (int i = 0; i < n; ++i) {
        int num;
        std::cin >> num;
        nums.push_back(num);
    }
    solution(n, nums);
}
